apps/sms/services/js/service_manager.js

```diff
@@ -0,0 +1,239 @@
+/* global App,
+          ConversationClient,
+          ConversationService,
+          MessagingClient,
+          MessagingService,
+          MozMobileConnectionsClient,
+          MozMobileConnectionsShim,
+          MozSettingsShim
+*/
+
+(function(exports) {
+  'use strict';
+
+  const debug = 1 ?
+    (arg1, ...args) => console.log(`[ServiceManager] ${arg1}`, ...args):
+    () => {};
+
+  const priv = {
+    upgradeKey: Symbol('upgradeKey')
+  };
+
+  const UPGRADE_KEY = 'upgrade-key';
+
+  // Allow old clients to gracefully finish all pending operations before
+  // we destroy them.
+  var ConversationServiceInitializer = {
```

Comment:
It seems like Initializers have initialize/upgrade methods but doing slightly different things. I'm thinking of spiting the service/shim into different prototype or mixin, not sure if it could reduce the duplicate part.